lib/devise-security/validators/password_complexity_validator.rb

```diff
@@ -1,37 +1,56 @@
 # frozen_string_literal: true
 
-# Password complexity validator
+# Password Complexity Validator
+#
 # Options:
-# - digit:  minimum number of digits in the validated string
-# - digits: minimum number of digits in the validated string
-# - lower:  minimum number of lower-case letters in the validated string
-# - symbol: minimum number of punctuation characters or symbols in the validated string
-# - symbols: minimum number of punctuation characters or symbols in the validated string
-# - upper:  minimum number of upper-case letters in the validated string
+# - `digit | digits`:  minimum number of digits in the validated string. Uses
+#   the `digit` localization key.
+# - `lower`:  minimum number of lower-case letters in the validated string
+# - `symbol | symbols`: minimum number of punctuation characters or symbols in
+#   the validated string. Uses the `symbol` localization key.
+# - `upper`:  minimum number of upper-case letters in the validated string
 class DeviseSecurity::PasswordComplexityValidator < ActiveModel::EachValidator
-  PATTERNS = {
-    digit: /\p{Digit}/,
-    digits: /\p{Digit}/,
-    lower: /\p{Lower}/,
-    symbol: /\p{Punct}|\p{S}/,
-    symbols: /\p{Punct}|\p{S}/,
-    upper: /\p{Upper}/
-  }.freeze
+  # A Hash of the possible valid patterns that can be checked against. The keys
+  # for this Hash are singular symbols corresponding to entries in the
+  # localization files. Override or redefine this method if you want to include
+  # custom patterns (e.g., `letter: /\p{Alpha}/` for all letters).
+  #
+  # @return [Hash<Symbol,Regexp>]
+  def patterns
+    {
+      digit: /\p{Digit}/,
+      lower: /\p{Lower}/,
+      symbol: /\p{Punct}|\p{S}/,
+      upper: /\p{Upper}/
+    }
+  end
 
-  def validate_each(record, attribute, value)
-    return unless value.present?
+  # @param record [ActiveModel::Model]
+  # @param attribute [Symbol]
+  # @param password [String]
+  def validate_each(record, attribute, password)
+    return if password.blank?
 
-    active_pattern_keys.each do |key|
-      minimum = [0, options[key].to_i].max
-      pattern = Regexp.new PATTERNS[key]
+    active_patterns.each do |pattern_name, minimum|
+      next if password.scan(patterns[pattern_name]).size >= minimum
 
-      unless (value || '').scan(pattern).size >= minimum
-        record.errors.add attribute, :"password_complexity.#{key}", count: minimum
-      end
+      record.errors.add(
+        attribute,
+        :"password_complexity.#{pattern_name}",
+        count: minimum
+      )
     end
   end
 
-  def active_pattern_keys
-    options.keys & PATTERNS.keys
+  # A Hash of the valid patterns that have been enabled. These are normalized to
+  # have singular key names and positive limits. Negative or zero limits are
+  # ignored.
+  #
+  # @return [Hash<Symbol,Integer>]
+  def active_patterns
+    options.transform_keys { |pattern| pattern.to_s.singularize.to_sym } # singularize keys
```

Comment:
I think you can make this a lot cleaner with a regular loop and a few extra lines of code instead of chained pretty methods. It should also condense this from 4 loops into 1 (not that performance is a huge deal with a small constant size array...)

```ruby
    {}.tap do |active_patterns|
      options.each do |option, value|
        normalized_option = option.to_s.singularize.to_sym
        next unless patterns.key?(normalized_option)

        transformed_value = [0, value.to_i].max # I think it's possible we could just do value.to_i.positive? on the next line
        active_patterns[normalized_option] = transformed_value if transformed_value.positive?
      end
    end
```

Reply:
I took this idea and used it to simplify the validator even more